cmd/nodecmd/join.go

```diff
@@ -0,0 +1,498 @@
+	}
+	return d, nil
+}
+
+func checkNodeIsBootstrapped(clusterName string, printOutput bool) (bool, error) {
+	ux.Logger.PrintToUser("Checking if node is bootstrapped to Primary Network ...")
+	if err := app.CreateAnsibleStatusFile(app.GetBootstrappedJSONFile()); err != nil {
+		return false, err
+	}
+	if err := ansible.RunAnsiblePlaybookCheckBootstrapped(app.GetAnsibleDir(), app.GetBootstrappedJSONFile(), app.GetAnsibleInventoryPath(clusterName)); err != nil {
+		return false, err
+	}
+	isBootstrapped, err := parseBootstrappedOutput(app.GetBootstrappedJSONFile(), printOutput)
+	if err != nil {
+		return false, err
+	}
+	if err := app.RemoveAnsibleStatusDir(); err != nil {
+		return false, err
+	}
+	if isBootstrapped {
+		return true, nil
+	}
+	return false, nil
+}
+
+func getClusterNodeID(clusterName string) (string, error) {
+	ux.Logger.PrintToUser("Getting node id ...")
+	if err := app.CreateAnsibleStatusFile(app.GetNodeIDJSONFile()); err != nil {
+		return "", err
+	}
+	if err := ansible.RunAnsiblePlaybookGetNodeID(app.GetAnsibleDir(), app.GetNodeIDJSONFile(), app.GetAnsibleInventoryPath(clusterName)); err != nil {
+		return "", err
+	}
+	nodeID, err := parseNodeIDOutput(app.GetNodeIDJSONFile())
+	if err != nil {
+		return "", err
+	}
+	if err = app.RemoveAnsibleStatusDir(); err != nil {
+		return "", err
+	}
+	return nodeID, err
+}
+
+func getNodeSubnetSyncStatus(blockchainID, clusterName string, printOutput bool) (bool, error) {
+	ux.Logger.PrintToUser("Checking if node is synced to subnet ...")
+	if err := app.CreateAnsibleStatusFile(app.GetSubnetSyncJSONFile()); err != nil {
+		return false, err
+	}
+	if err := ansible.RunAnsiblePlaybookSubnetSyncStatus(app.GetAnsibleDir(), app.GetSubnetSyncJSONFile(), blockchainID, app.GetAnsibleInventoryPath(clusterName)); err != nil {
+		return false, err
+	}
+	subnetSyncStatus, err := parseSubnetSyncOutput(app.GetSubnetSyncJSONFile(), printOutput)
+	if err != nil {
+		return false, err
+	}
+	if err = app.RemoveAnsibleStatusDir(); err != nil {
+		return false, err
+	}
+	if subnetSyncStatus == status.Syncing.String() {
+		return true, nil
+	} else if subnetSyncStatus == status.Validating.String() {
+		return false, errors.New("node is already a subnet validator")
+	}
+	return false, nil
+}
+
+// checkNodeIsPrimaryNetworkValidator only returns err if node is already a Primary Network validator
+func checkNodeIsPrimaryNetworkValidator(nodeID ids.NodeID, network models.Network) error {
+	isValidator, err := subnet.IsSubnetValidator(ids.Empty, nodeID, network)
+	if err != nil {
+		ux.Logger.PrintToUser("failed to check if node is a validator on Primary Network: %s", err)
+	} else if isValidator {
+		return fmt.Errorf("node %s is already a validator on Primary Network", nodeID)
+	}
+	return nil
+}
+
+func addNodeAsPrimaryNetworkValidator(nodeID ids.NodeID, network models.Network) (bool, error) {
+	if err := checkNodeIsPrimaryNetworkValidator(nodeID, network); err == nil {
```

Comment:
err is ignored, please make checkNodeIsPrimaryNetworkValidator to return (bool, error) to differentiate check failure from validtion status, then use the value to either return quickly cause it is already a validator, or aftertwards add it as validator